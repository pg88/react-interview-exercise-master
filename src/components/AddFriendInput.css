:local(.addFriendInput) {
    border-radius: 0;
    border-color: #ABAAAA;
    border-left: 0;
    border-right: 0;
}

:local(.addFriendsButton) {
    width: 100%;
    padding: 4px;
    color: white;
    cursor: pointer;
    border-radius: 5px;
    transition: all 1s;
    background: #597aff;
    border: 1px solid #415BB7;
    -webkit-transition: all 1s; /* Safari */

}
:local(.addFriendsButton:hover:not([disabled])) {
    background: #415BB7;
    border: 1px solid #597aff;
}