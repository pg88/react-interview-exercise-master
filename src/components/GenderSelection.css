:local(.genderContainer) {
    width: 100%;
    display: block;
}
:local(.genderInputs) {
    display: inline-block;
    margin: 5px;
}