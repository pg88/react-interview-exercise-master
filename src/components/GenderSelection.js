import React, { Component } from 'react';
import { connect } from 'react-redux';
import styles from './GenderSelection.css';
import { OTHER, GENDER } from '../constants/ActionTypes';


class GenderSelection extends Component {
    constructor(props, context) {
        super(props, context);
        this.state = {
            gender: OTHER
        }
    }

    selectGender(e) {
        const selectedGender = e.target.value;
        this.setState({ gender: selectedGender });
        this.props.onSelectGender(selectedGender)
    }

    displayGenders() {
        return GENDER.map((value, key) => {
            return (<div key={ value } className={ styles.genderInputs }>
                <input
                    type='radio'
                    name='gender'
                    value={ value }
                    tabIndex={ key }
                    onChange={ this.selectGender.bind(this) }
                /> { value }
            </div>)
        })
    }

    render () {
        return(
            <div className={ styles.genderContainer }>
                <div className={ styles.genderLabel }>
                    <span>Gender:</span>
                </div>
                { this.displayGenders() }
            </div>
        )
    }
}

export default GenderSelection