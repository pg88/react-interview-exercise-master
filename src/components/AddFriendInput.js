import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import styles from './AddFriendInput.css';
import GenderSelection from './GenderSelection'
import { OTHER } from '../constants/ActionTypes';

class AddFriendInput extends Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            name: this.props.name || '',
            gender: OTHER,
            error: null
        };
    }

    handleChange(e) {
        this.setState({ name: e.target.value });
    }

    validate() {
        const name = this.state.name;
        if (name.trim().length === 0) {
            this.setState({
                error: true,
            });
            return false
        }
        return true;
    }

    handleSubmit(e) {
        const name = e.target.value.trim();
        if (e.which === 13 && this.validate()) {
           this.addNewFriend(name);
        }
    }
    addNewFriend(name) {
        this.props.addFriend(name, this.state.gender);
        this.setState({ name: '' });
    }
    handleForm() {
        return this.validate() ? this.addNewFriend(this.state.name) : false;
    }
    setGender (gender) {
        this.setState({ gender: gender });
    }

    render() {
        return (
            <div>
                <input
                    type="text"
                    autoFocus="true"
                    className={ classnames('form-control', styles.addFriendInput) }
                    placeholder="Type the name of a friend"
                    value={ this.state.name }
                    onChange={ this.handleChange.bind(this) }
                    onKeyDown={ this.handleSubmit.bind(this) }/>
                <GenderSelection onSelectGender={ this.setGender.bind(this) }/>
                <button className={ styles.addFriendsButton } disabled={ !this.state.name } onClick={ this.handleForm.bind(this) }>
                    ADD FRIEND
                </button>
            </div>
        );
    }

}

AddFriendInput.propTypes = {
    addFriend: PropTypes.func.isRequired
};
export default AddFriendInput;
