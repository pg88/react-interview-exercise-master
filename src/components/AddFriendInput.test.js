import React from 'react';
import { shallow } from 'enzyme';
import AddFriendInput from './AddFriendInput';
import GenderSelection from './GenderSelection';
import { MALE } from '../constants/ActionTypes';



describe('AddFriendInput', () => {
    const propsData = {
        addFriend: jest.fn()
    };
    it('should render correctly', () => {
        expect(shallow(<AddFriendInput {...propsData} />)).toMatchSnapshot();
    });
    /*it('should try to add empty data for', () => {
        const component = shallow(<AddFriendInput {...propsData} />);
        component.setState({ name: '', gender: MALE });
        component.find('button').simulate('click');
        expect(component.state().error).not.toBeNull();
    });
    it('should have the friend properties', () => {
        const component = shallow(<AddFriendInput {...propsData} />);
        expect(component.find("input[type='radio']").length).toBe(GENDER.length);
        expect(component.find("input[type='text']").length).toBe(1);
        expect(component.find('button').length).toBe(1);
    })*/
});
