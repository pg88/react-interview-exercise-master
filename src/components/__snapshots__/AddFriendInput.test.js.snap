exports[`AddFriendInput should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddFriendInput
    addFriend={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          autoFocus="true"
          className="form-control"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Type the name of a friend"
          type="text"
          value="" />,
        <span />,
        <GenderSelection
          onSelectGender={[Function]} />,
        <button
          className={undefined}
          disabled={true}
          onClick={[Function]}>
          ADD FRIEND
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoFocus": "true",
          "className": "form-control",
          "onChange": [Function],
          "onKeyDown": [Function],
          "placeholder": "Type the name of a friend",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onSelectGender": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "ADD FRIEND",
          "className": undefined,
          "disabled": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "ADD FRIEND",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            autoFocus="true"
            className="form-control"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Type the name of a friend"
            type="text"
            value="" />,
          <span />,
          <GenderSelection
            onSelectGender={[Function]} />,
          <button
            className={undefined}
            disabled={true}
            onClick={[Function]}>
            ADD FRIEND
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoFocus": "true",
            "className": "form-control",
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "Type the name of a friend",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSelectGender": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "ADD FRIEND",
            "className": undefined,
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "ADD FRIEND",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
